mllib/src/main/scala/org/apache/spark/mllib/recommendation/MatrixFactorizationModel.scala

```diff
@@ -89,10 +89,14 @@ class MatrixFactorizationModel(
    */
   def predict(usersProducts: RDD[(Int, Int)]): RDD[Rating] = {
     val users = userFeatures.join(usersProducts).map {
-      case (user, (uFeatures, product)) => (product, (user, uFeatures))
+      case (user, (uFeatures, product)) => ((user, product), uFeatures)
     }
-    users.join(productFeatures).map {
-      case (product, ((user, uFeatures), pFeatures)) =>
+    val productUsers = usersProducts.map(up => (up._2, up._1))
+    val products = productFeatures.join(productUsers).map {
```

Comment:
.. but this creates an entire other `join` to compute. This is going to be significantly slower, right? How does this relate to the JIRA though -- this is a more expensive version of a simple repartitioning, which would also solve it, but the point seemed to be to avoid that.